core/common/src/Buffer.kt

```diff
@@ -20,396 +20,679 @@
-  ): Buffer
+    val data = segment.data
+    val s = data[pos++] and 0xff shl 8 or (data[pos++] and 0xff)
+    size -= 2L
+
+    if (pos == limit) {
+      head = segment.pop()
+      SegmentPool.recycle(segment)
+    } else {
+      segment.pos = pos
+    }
+
+    return s.toShort()
+  }
+
+  override fun readInt(): Int {
+    require(4)
+
+    val segment = head!!
+    var pos = segment.pos
+    val limit = segment.limit
+
+    // If the int is split across multiple segments, delegate to readByte().
+    if (limit - pos < 4L) {
+      return (
+              readByte() and 0xff shl 24
+                      or (readByte() and 0xff shl 16)
+                      or (readByte() and 0xff shl 8)
+                      or (readByte() and 0xff)
+              )
+    }
+
+    val data = segment.data
+    val i = (
+            data[pos++] and 0xff shl 24
+                    or (data[pos++] and 0xff shl 16)
+                    or (data[pos++] and 0xff shl 8)
+                    or (data[pos++] and 0xff)
+            )
+    size -= 4L
+
+    if (pos == limit) {
+      head = segment.pop()
+      SegmentPool.recycle(segment)
+    } else {
+      segment.pos = pos
+    }
+
+    return i
+  }
+
+  override fun readLong(): Long {
+    require(8)
+
+    val segment = head!!
+    var pos = segment.pos
+    val limit = segment.limit
+
+    // If the long is split across multiple segments, delegate to readInt().
+    if (limit - pos < 8L) {
+      return (
+              readInt() and 0xffffffffL shl 32
+                      or (readInt() and 0xffffffffL)
+              )
+    }
+
+    val data = segment.data
+    val v = (
+            data[pos++] and 0xffL shl 56
+                    or (data[pos++] and 0xffL shl 48)
+                    or (data[pos++] and 0xffL shl 40)
+                    or (data[pos++] and 0xffL shl 32)
+                    or (data[pos++] and 0xffL shl 24)
+                    or (data[pos++] and 0xffL shl 16)
+                    or (data[pos++] and 0xffL shl 8)
+                    or (data[pos++] and 0xffL)
+            )
+    size -= 8L
+
+    if (pos == limit) {
+      head = segment.pop()
+      SegmentPool.recycle(segment)
+    } else {
+      segment.pos = pos
+    }
+
+    return v
+  }
+
+  /**
+   * This method does not affect the buffer's content as there is no upstream to write data to.
+   */
+  @InternalIoApi
+  override fun hintEmit(): Unit = Unit
 
   /**
-   * Overload of [copyTo] with byteCount = size - offset, work around for
-   *  https://youtrack.jetbrains.com/issue/KT-30847
+   * This method does not affect the buffer's content as there is no upstream to write data to.
    */
-  fun copyTo(
+  override fun emit(): Unit = Unit
+
+  /**
+   * This method does not affect the buffer's content as there is no upstream to write data to.
+   */
+  override fun flush(): Unit = Unit
+
+  /**
+   * Copy bytes from this buffer's subrange starting at [startIndex] and ending at [endIndex], to [out] buffer.
```

Comment:
Does this operation not consume bytes from the current buffer?
If so, can it be specified explicitly?